rpc/dial.go

```diff
@@ -93,90 +108,161 @@ func dial(
 	dOpts *dialOptions,
 	tryLocal bool,
 ) (ClientConn, bool, error) {
+	if ctx.Err() != nil {
+		return nil, false, ctx.Err()
+	}
+
 	var isJustDomain bool
-	if strings.HasPrefix(address, "unix://") {
+	switch {
+	case strings.HasPrefix(address, "unix://"):
 		dOpts.mdnsOptions.Disable = true
 		dOpts.webrtcOpts.Disable = true
 		dOpts.insecure = true
 		dOpts.disableDirect = false
-	} else {
-		if strings.ContainsRune(address, ':') {
-			isJustDomain = false
-		} else {
-			isJustDomain = net.ParseIP(address) == nil
-		}
+	case strings.ContainsRune(address, ':'):
+		isJustDomain = false
+	default:
+		isJustDomain = net.ParseIP(address) == nil
 	}
 
+	var (
+		wg                          sync.WaitGroup
+		dialCh                      = make(chan dialResult)
+		ctxParallel, cancelParallel = context.WithCancel(ctx)
+	)
+	defer cancelParallel()
 	if !dOpts.mdnsOptions.Disable && tryLocal && isJustDomain {
-		conn, cached, err := dialMulticastDNS(ctx, address, logger, dOpts)
-		if err != nil {
-			logger.Warnw("error dialing with mDNS; falling back to other methods", "error", err)
-		}
-
-		if conn != nil {
-			return conn, cached, err
-		}
+		dOptsCopy := *dOpts
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			conn, cached, err := dialMulticastDNS(ctxParallel, address, logger, &dOptsCopy)
+			switch {
+			case err != nil:
+				dialCh <- dialResult{err: err}
+			case conn != nil:
+				// TODO(docs): how can we get a `nil` connection when there is no error?
+				dialCh <- dialResult{conn: conn, cached: cached, usedMDNS: true}
```

Comment:
from what I can tell, there are some code paths where mDNS does not find a host but there is no associated error. we really should throw an error in this case instead... but I'd prefer to do this on a follow-up PR